### ISSUE TITLE
Using mpdf in phar package leads to weird errors

### ISSUE THREAD

**Issue by sandreas** [mpdf/mpdf#1504]
### I found this bug
While `mpdf` works fine, when using the source code / script files, running the same scripts packaged as phar results in weird errors:

```
PHP Warning:  Undefined array key 4279 in phar:///home/sandreas/projects/lblmk/dist/lblmk.phar/vendor/mpdf/mpdf/src/TTFontFile.php on line 4536
PHP Stack trace:
PHP   1. {main}() /home/sandreas/projects/lblmk/dist/lblmk.phar:0
PHP   2. require() /home/sandreas/projects/lblmk/dist/lblmk.phar:12
PHP   3. Mpdf\Mpdf->Output($name = '', $dest = 'S') phar:///home/sandreas/projects/lblmk/dist/lblmk.phar/src/index.php:13
PHP   4. Mpdf\Mpdf->Close() phar:///home/sandreas/projects/lblmk/dist/lblmk.phar/vendor/mpdf/mpdf/src/Mpdf.php:9484
PHP   5. Mpdf\Mpdf->_enddoc() phar:///home/sandreas/projects/lblmk/dist/lblmk.phar/vendor/mpdf/mpdf/src/Mpdf.php:2048
PHP   6. Mpdf\Writer\ResourceWriter->writeResources() phar:///home/sandreas/projects/lblmk/dist/lblmk.phar/vendor/mpdf/mpdf/src/Mpdf.php:10001
PHP   7. Mpdf\Writer\FontWriter->writeFonts() phar:///home/sandreas/projects/lblmk/dist/lblmk.phar/vendor/mpdf/mpdf/src/Writer/ResourceWriter.php:113
PHP   8. Mpdf\TTFontFile->makeSubset($file = 'phar:///home/sandreas/projects/lblmk/dist/lblmk.phar/vendor/mpdf/mpdf/src/Config/../../ttfonts/DejaVuSerifCondensed.ttf', $subset = [32 => 32, 33 => 33, 34 => 34, 35 => 35, 36 => 36, 37 => 37, 38 => 38, 39 => 39, 40 => 40, 41 => 41, 42 => 42, 43 => 43, 44 => 44, 45 => 45, 46 => 46, 47 => 47, 48 => 48, 49 => 49, 50 => 50, 51 => 51, 52 => 52, 53 => 53, 54 => 54, 55 => 55, 56 => 56, 57 => 57, 58 => 58, 59 => 59, 60 => 60, 61 => 61, 62 => 62, 63 => 63, 64 => 64, 65 => 65, 66 => 66, 67 => 67, 68 => 68, 69 => 69, 70 => 70, 71 => 71, 72 => 72, 73 => 73, 74 => 74, 75 => 75, 76 => 76, 77 => 77, 78 => 78, 79 => 79, 80 => 80, 81 => 81, 82 => 82, 83 => 83, 84 => 84, 85 => 85, 86 => 86, 87 => 87, 88 => 88, 89 => 89, 90 => 90, 91 => 91, 92 => 92, 93 => 93, 94 => 94, 95 => 95, 96 => 96, 97 => 97, 98 => 98, 99 => 99, 100 => 100, 101 => 101, 102 => 102, 103 => 103, 104 => 104, 105 => 105, 106 => 106, 107 => 107, 108 => 108, 109 => 109, 110 => 110, 111 => 111, 112 => 112, 113 => 113, 114 => 114, 115 => 115, 116 => 116, 117 => 117, 118 => 118, 119 => 119, 120 => 120, 121 => 121, 122 => 122, 123 => 123, 124 => 124, 125 => 125, 126 => 126, 127 => 127], $TTCfontID = 0, $debug = FALSE, $useOTL = FALSE) phar:///home/sandreas/projects/lblmk/dist/lblmk.phar/vendor/mpdf/mpdf/src/Writer/FontWriter.php:286
PHP   9. Mpdf\TTFontFile->getHMTX($numberOfHMetrics = 3438, $numGlyphs = 3438, $glyphToChar = [0 => [0 => 0, 1 => 65535], 3 => [0 => 32], 4 => [0 => 33], 5 => [0 => 34], 6 => [0 => 35], 7 => [0 => 36], 8 => [0 => 37], 9 => [0 => 38], 10 => [0 => 39], 11 => [0 => 40], 12 => [0 => 41], 13 => [0 => 42], 14 => [0 => 43], 15 => [0 => 44], 16 => [0 => 45], 17 => [0 => 46], 18 => [0 => 47], 19 => [0 => 48], 20 => [0 => 49], 21 => [0 => 50], 22 => [0 => 51], 23 => [0 => 52], 24 => [0 => 53], 25 => [0 => 54], 26 => [0 => 55], 27 => [0 => 56], 28 => [0 => 57], 29 => [0 => 58], 30 => [0 => 59], 31 => [0 => 60], 32 => [0 => 61], 33 => [0 => 62], 34 => [0 => 63], 35 => [0 => 64], 36 => [0 => 65], 37 => [0 => 66], 38 => [0 => 67], 39 => [0 => 68], 40 => [0 => 69], 41 => [0 => 70], 42 => [0 => 71], 43 => [0 => 72], 44 => [0 => 73], 45 => [0 => 74], 46 => [0 => 75], 47 => [0 => 76], 48 => [0 => 77], 49 => [0 => 78], 50 => [0 => 79], 51 => [0 => 80], 52 => [0 => 81], 53 => [0 => 82], 54 => [0 => 83], 55 => [0 => 84], 56 => [0 => 85], 57 => [0 => 86], 58 => [0 => 87], 59 => [0 => 88], 60 => [0 => 89], 61 => [0 => 90], 62 => [0 => 91], 63 => [0 => 92], 64 => [0 => 93], 65 => [0 => 94], 66 => [0 => 95], 67 => [0 => 96], 68 => [0 => 97], 69 => [0 => 98], 70 => [0 => 99], 71 => [0 => 100], 72 => [0 => 101], 73 => [0 => 102], 74 => [0 => 103], 75 => [0 => 104], 76 => [0 => 105], 77 => [0 => 106], 78 => [0 => 107], 79 => [0 => 108], 80 => [0 => 109], 81 => [0 => 110], 82 => [0 => 111], 83 => [0 => 112], 84 => [0 => 113], 85 => [0 => 114], 86 => [0 => 115], 87 => [0 => 116], 88 => [0 => 117], 89 => [0 => 118], 90 => [0 => 119], 91 => [0 => 120], 92 => [0 => 121], 93 => [0 => 122], 94 => [0 => 123], 95 => [0 => 124], 96 => [0 => 125], 97 => [0 => 126], 98 => [0 => 160], 99 => [0 => 161], 100 => [0 => 162], 101 => [0 => 163], 102 => [0 => 164], 103 => [0 => 165], 104 => [0 => 166], 105 => [0 => 167], 106 => [0 => 168], 107 => [0 => 169], 108 => [0 => 170], 109 => [0 => 171], 110 => [0 => 172], 111 => [0 => 173], 112 => [0 => 174], 113 => [0 => 175], 114 => [0 => 176], 115 => [0 => 177], 116 => [0 => 178], 117 => [0 => 179], 118 => [0 => 180], 119 => [0 => 181], 120 => [0 => 182], 121 => [0 => 183], 122 => [0 => 184], 123 => [0 => 185], 124 => [0 => 186], 125 => [0 => 187], 126 => [0 => 188], 127 => [0 => 189], 128 => [0 => 190], 129 => [0 => 191], ...], $scale = 1)
```

I'm using the box-project (https://github.com/box-project/box) to create a phar and was not able to track down the error to a specific part of code. Below is what I found out, I also attached a zip file containing a full project template including latest `box` release to reproduce the error as easy as possible:

[lblmk.zip](https://github.com/mpdf/mpdf/files/7164820/lblmk.zip)

*project template instructions* 
- extract the zip file into an extra project folder (e.g. `lblmk`)
- `cd lblmk`
- `composer install`
- `./run-source` will produce an `output-source.pdf` in the same directory
- `./run-phar` will compile the phar (if not present), run it and spit out a vast amount of errors instead of creating `output-phar.pdf`

*what I found out so far*
- The `phar` file can be debugged like a normal script in PHPStorm
- The index of a glyph is not defined when using the phar
  - https://github.com/mpdf/mpdf/blob/8107f1103639c14726e865298eb9a00eac23ab2a/src/TTFontFile.php#L4536
  - https://github.com/mpdf/mpdf/blob/8107f1103639c14726e865298eb9a00eac23ab2a/src/TTFontFile.php#L4627 
- If I add an isset to prevent the invalid array index access, the error is `Path cannot be empty`, which may occur due to an empty string as filename for `file_get_contents` somewhere

### This is mPDF and PHP version and environment (server/fpm/cli etc) I am using

- mpdf `8.0.13`
- PHP `8.0.10` cli

### This is the PHP code snippet I use

```
<?php
declare(strict_types=1);

require __DIR__.'/../vendor/autoload.php';

use Mpdf\Mpdf;
use Mpdf\Output\Destination;
$html = "<html><body>hello</body></html>";
$mpdf = new Mpdf(['tempDir' => sys_get_temp_dir()]);
$mpdf->img_dpi = 300;
$mpdf->WriteHTML($html);

file_put_contents($argv[1],  $mpdf->Output('', Destination::STRING_RETURN));
```

### This is the HTML/CSS code snippet I use

```
<html><body>hello</body></html>
```

**Comment by finwe** [comment]
Tagging as an enhancement as packing mPDF into a phar is not a _supported method of use_ of mPDF.

**Comment by sandreas** [comment]
> Tagging as an enhancement as packing mPDF into a phar is not a supported method of use of mPDF.

Ok, maybe I can help out. Don't expect too much ;)

**Comment by sandreas** [comment]
Short update - I tracked down the problem to here: https://github.com/mpdf/mpdf/blob/8107f1103639c14726e865298eb9a00eac23ab2a/src/TTFontFile.php#L4527

At some point, `$start` vastly increases (e.g. from `30208` to `255444`), enters the `if (($numberOfHMetrics * 4) < $this->maxStrLenRead) {` condition and from there `$data` is different (`13752 bytes` in source vs. `8192 bytes` in phar, although in both cases the `$numberOfHMetrics * 4 === 13752` ):

```php
$numberOfHMetrics = 3438;
$this->maxStrLenRead = 200000;
$start = 255444;

if (($numberOfHMetrics * 4) < $this->maxStrLenRead) {

  $data = $this->get_chunk($start, $numberOfHMetrics * 4);
  // phar: strlen($data) === 8192
  // source: strlen($data) === 13752

  $arr = unpack('n*', $data);
}
```

It may be a caching issue, if I find out more, I'll post an update.

**Comment by sandreas** [comment]
Ok, I think, I found it.
https://github.com/mpdf/mpdf/blob/8107f1103639c14726e865298eb9a00eac23ab2a/src/TTFontFile.php#L521

fread can be used, to read binary data, but:

> if the stream is read buffered and it does not represent a plain file, at most one read of up to a number of bytes equal to the chunk size (usually 8192) is made; depending on the previously buffered data, the size of the returned data may be larger than the chunk size. 
Source: https://www.php.net/manual/en/function.fread.php

So a fix would be:

```php
	function get_chunk($pos, $length)
	{

		fseek($this->fh, $pos);
		if ($length < 1) {
			return '';
		}

        $data = (fread($this->fh, $length));
        $dataLen = strlen($data);
        while($dataLen < $length && !feof($this->fh)) {
            $data .= fread($this->fh, $length - $dataLen);
            $dataLen = strlen($data);
        }
        return $data;
    }
```

Would you like me to create a PR, or would you like to do further investigation?

**Comment by finwe** [comment]
Thanks for the investigation. I'd like to understand why this happens only when used in PHAR. Any idea?

**Comment by sandreas** [comment]
Eehr, I thought I already wrote it ;-)

> if the stream is read buffered and it does not represent a plain file, at most one read of up to a number of bytes equal to the chunk size (usually 8192) is made; depending on the previously buffered data, the size of the returned data may be larger than the chunk size.
> Source: https://www.php.net/manual/en/function.fread.php

Explanation: PHAR is a compressed format, not a plain file. So if it tries to load a file or font via `fread`, it buffers this file and is only able to read the number of bytes equal to the chunk size. Most of the time, the filesize is below 8192 bytes and it works flawlessly,  but some files are bigger than 8192 bytes and then it fails. This is why it was a kind of weird behaviour...

**Comment by finwe** [comment]
Sure. Fine, feel free to prepare a PR, I'd be happy to incorporate it.

**Comment by sandreas** [comment]
Should be fixed in #1506 

Comments are welcome. Thanks for the nice and clean communication.
